flybywiresim/a32nx
# feat(efb): Flypadosv3 checklists

Fixes #4443

## Summary of Changes

Added support for checklists in EFB - following up on issue #4443 
- manual checked items that can be checked / unchecked by clicking on the item 
- automtically checked items based on the state of the aircraft (state could be overwritten by double clicking)
- setting a checklist complete and advancing to the next checklist in the sequence
- resetting already complete checklists

## Screenshots (if necessary)

### Checklist in its inital state (e.g. "Before start" checklist)

![01_Checklist_Start](https://user-images.githubusercontent.com/768655/143836242-3430935b-0a27-4285-96e5-b33b9aad8a7d.png)
- Before start checklist active
- No manual item checked yet - toggle check state by single clicking on them
- One automatic item already checked based on the state of the A/C (the automatic checked items are marked with *)

### Checklist with a few items already checked

![02_Checklist_FewItemsChecked](https://user-images.githubusercontent.com/768655/143836550-75a59ed2-261c-4576-bb78-f092dcef7f96.png)
- Few manual items are selected (items in green)

### Checklist with an overwritten automatic checked item

![03_Checklist_AutomaticResultOverwritten](https://user-images.githubusercontent.com/768655/143836879-a05d62ef-f12a-4724-8a71-1ccc6b17597e.png)
- One automatically checked item has been overwritten (ADIRS) and marked checked (colored in blue to distinguish from the manual items) - toggle the overwrite by double clicking onto this item

### Checklist with all items checked - not yet completed

![04_Checklist_Complete](https://user-images.githubusercontent.com/768655/143837246-fa10800f-51b3-47d1-9664-4873e88b21ce.png)
- All checklist items checked
- Checklist can be completed by clicking on the "Complete checklist" button

### One checklist completed

![05_Checklist_AfterCompleteSelected](https://user-images.githubusercontent.com/768655/143837620-a0d7e090-7400-4104-95c1-0a2b40cb370c.png)
- After "Checklist complete" has been selected, the next checklist is automatically displayed
- After a checklist is complete, the automatically checked items are no longer evaluated

### Checklist reset

![06_Checklist_ChecklistReset](https://user-images.githubusercontent.com/768655/143837772-115b1d21-5f56-4028-bb18-4f85b7f3535f.png)
- After a checklist has been set to complete, it can be reset by selecting "Reset checklist" in case you like to go over the items again.

## References

Using the checklist provided in #4443 

## Additional context

As discussed with EricSharp, I am creating this as a pull request for the flypadosv3 branch.

Discord username (if different from GitHub): geholger#9751

## Testing instructions

1. Open EFB after aircraft is started
2. Select Checklist from the Toolbar (the checkmark)
3. Select a checklist
4. Single click on a manual checklist items 
 - item will toggle between white (not checked) and green (checked)
5. Change condition for automatic checklist items, e.g. rudder trim, beacon, TCAS, ...) to toggle the state for those items
 - checklist item should be shown in white (no checked) resp. green (checked) if all conditions are ok
6. Overwrite a checked state for an unchecked automatic item 
 - e.g. set rudder trim set to non-zero --> checklist item should then be shown in white 
 - double clicking on this item --> item should be marked checked but shown in blue
7. Complete a checklist 
 - list should be marked green in the tabs
 - next checklist should be displayed automatically
 8. Reset a completed checklist
 - complete a checklist  / see 7,
 - select checklist again and click on "Reset checklist"
 - all manual items should be shown on white
 - automatic items show the current state of the condition for the item (white or green)
 - no automatic items are overwritten


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
